design/design.md

```diff
@@ -0,0 +1,203 @@
+to our needs:
+
+* Since an HTTP request must be made to the CTS to fetch the job data, including the output files,
+  the SSS data sink must allow for arbitrary code. This
+  [limits the sinks](https://spark.apache.org/docs/latest/structured-streaming-programming-guide.html#output-sinks)
+  to the `forEach`   and `forEachBatch` sinks.
+* `forEach` code
+  [runs on the Spark cluster worker](https://spark.apache.org/docs/latest/api/java/org/apache/spark/sql/ForeachWriter.html#process-T-).
+  This means that the only the resources of that worker, not the entire cluster, can be used to
+  process the job output data. As such `forEach` is not an acceptable choice.
+* `forEachBatch` processes data on the driver, and so the driver code can submit full cluster
+  jobs for the CTS jobs in the batch. However,
+  [only one batch runs at a time](https://spark.apache.org/docs/latest/structured-streaming-programming-guide.html#triggers)
+  which means the processing time for the batch is equal to the processing time of largest
+  job output in the batch if the batch is parallelized. `forEachBatch` doesn't allow for
+  true parallelism - instead, from a performance perspective, it effectively picks the largest job
+  out of the N jobs in the batch, and sequentially runs those large jobs. Given the nature of
+  bioinformatics processing, job output processing times are expected to be highly variable,
+  with times ranging from seconds to hours or even days.
+* Parallelizing `forEachBatch` independently would case massive message duplication as each
+  SSS instance manages its own Kafka offsets.
+  
+As such, we will implement with a standard Kafka client like `python-kafka`.
+
+## Design
+
+For now the design will be focused solely on responding to CTS completed job events.
+
+* Parallelism is based on scaling multiple Docker containers, each running a sequential,
+  single threaded event loop with access to a Spark session.
+    * The parallelism is limited by the number of partitions in the Kafka topic.
+* Interaction with Kafka will be via the `python-kafka` client.
+* For the first iteration, the spark session will be recreated for each job.
+    * Future work could investigate whether there are performance improvements from
+      sharing the session and whether sharing may cause conflicts between importers.
+* For each Kafka message
+    * pull the CTS job based on the ID
+    * Map the job image name to the python importer code to run
+        * Mappings must be made available to the processor (see below)
+        * Image name -> code only, no tag or digest
+        * The importer code is responsible for dealing with processing changes based
+          on the image digest
+            * The image tag is not supplied as it is mutable - code should key off the
+              digest
+    * Create the Spark session
+        * Add a zip of dependency python code
+        * Only python supported
```

Comment:
This seems hacky...

Reply:
The zip of python code? That's just how spark distributes code to workers. It's essentially the same as distributing jars

Reply:
fair enough